### PR TITLE
Amortised inference for cell2location (hierarchical messenger guide + aggressive training)

### PR DESCRIPTION
**Pull request by vitkl**
- Hierarchical amortised inference messenger guide.
- Aggressive training plan.
- Mutual information computation between input data and amortised NN output.
- Modified version of FCLayers from scvi-tools with each layer starting rather than ending in dropout (aka applying dropout to NN input data).
- Improved posterior quantile computation and tests.
- Other experimental features.

### CODE REVIEW

File: cell2location/models/base/_pyro_mixin.py [hunk 1]
```diff
@@ -494,3 +497,103 @@ def plot_QC(self, summary_name: str = "means", use_n_obs: int = 1000):
...
+                loss = torch.Tensor([self.svi_amortised.step(*args, **kwargs)])
+            else:
+                self.aggressive_steps_counter = 0
+                # Do parameter update exclusively for non-amortised variables
+                loss = torch.Tensor([self.svi_nonamortised.step(*args, **kwargs)])
+        else:
+            # Do parameter update for both types of variables
+            loss = torch.Tensor([self.svi.step(*args, **kwargs)])
+
+        return {"loss": loss}
```
Comment (vitkl) [inline cell2location/models/base/_pyro_mixin.py:599]: I am wondering if this is the correct way to do aggressive updating of amortised variables as described here: https://arxiv.org/pdf/1901.05534.pdf

This approach heuristically sets the number of inner and outer loop training settings rather than using mutual information as described in the paper.

File: cell2location/distributions/AutoAmortisedNormalMessenger.py [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+            loc, scale = self._get_params(name, prior)
+
+        if name not in self.amortised_plate_sites["sites"].keys():
+            # if amortisation is not used for a particular site return MI=0
+            return 0
+
+        #### create tensors with useful numbers ####
+        one = torch.ones((), dtype=loc.dtype, device=loc.device)
+        two = torch.tensor(2, dtype=loc.dtype, device=loc.device)
+        pi = torch.tensor(3.14159265359, dtype=loc.dtype, device=loc.device)
```
Comment (yozhikoff) [inline cell2location/distributions/AutoAmortisedNormalMessenger.py:455]: I think you can just use `1`, `2`, and `torch.pi` here
Also why is `one` set to `()`?

File: cell2location/cluster_averages/cluster_averages.py [hunk 3]
```diff
@@ -53,29 +53,40 @@ def compute_cluster_averages(adata, labels, use_raw=True, layer=None):
...
-        raise ValueError("cluster_col is absent in adata_ref.obs or not unique")
+    if layer is not None:
+        x = adata.layers[layer]
+        var_names = adata.var_names
+    else:
+        if not use_raw:
+            x = adata.X
+            var_names = adata.var_names
+        else:
+            if not adata.raw:
```
[the remark this answers is not in the document]
Comment (vitkl) [reply cell2location/cluster_averages/cluster_averages.py:70]: Yes exactly

File: cell2location/models/simplified/_cell2location_v3_no_mg_module.py [hunk 4]
```diff
@@ -317,7 +317,7 @@ def forward(self, x_data, idx, batch_index):
         # =====================Gene-specific overdispersion ======================= #
         alpha_g_phi_hyp = pyro.sample(
             "alpha_g_phi_hyp",
-            dist.Gamma(self.alpha_g_phi_hyp_prior_alpha, self.alpha_g_phi_hyp_prior_beta),
+            dist.Gamma(self.ones * self.alpha_g_phi_hyp_prior_alpha, self.ones * self.alpha_g_phi_hyp_prior_beta),
```
[the remark this answers is not in the document]
Comment (vitkl) [reply cell2location/models/simplified/_cell2location_v3_no_mg_module.py:320]: `self.ones` has shape = (1, 1) - I use it to make sure that tensors become at least 2D

File: cell2location/distributions/AutoAmortisedNormalMessenger.py (hunk 2 above)
Comment (vitkl) [reply cell2location/distributions/AutoAmortisedNormalMessenger.py:455]: All 3 should be scalar tensors, have the same dtype and device as loc. For some reason, when using torch.pi, I got errors with device location.
